lua/entities/gmod_wire_expression2/core/selfaware.lua

```diff
@@ -91,19 +91,88 @@ end)
 /******************************************************************************/
 -- Name functions
 
--- Set the name of the E2 itself
-e2function void setName( string name )
-	local e = self.entity
-	if (e.name == name) then return end
-	if (name == "generic" or name == "") then
+local SetNameHelper = function(ent, name)
+	-- NOTE: This function might not work as expected when called on other (non-E2) Wire entities.
+	if ent.name == name then return end -- Prevent it from running again if the name is unchanged.
+	if name == "generic" or name == "" then
 		name = "generic"
-		e.WireDebugName = "Expression 2"
+		ent.WireDebugName = "Expression 2"
 	else
-		e.WireDebugName = "E2 - " .. name
+		ent.WireDebugName = "E2 - " .. name
+	end
+	ent.name = name
+	ent:SetNWString("name", name)
+	ent:SetOverlayText(name)
+end
+
+local IsValid = IsValid
+local string = string
+local string_find = string.find
+
+local SetComponentNameHelper = function(ent, componentName)
+	if ent.wireName == componentName or string_find(componentName, "[\n\r\"]") ~= nil then return end
+	ent.wireName = componentName
+	ent:SetNWString("WireName", componentName)
+	duplicator.StoreEntityModifier(ent, "WireName", { name = componentName })
+end
+
+local GetComponentNameHelper = function(ent)
+	-- TODO: Check if it is wire-based entity! Possibly use scripted_ents library to figure out the base entity through the class name?
+	-- ISSUE: https://github.com/wiremod/wire/issues/1367
+	return IsValid(ent) and type(ent) == "Entity" and ent:GetNWString("WireName", ent.PrintName) or "" -- Prevent it from running on players/vehicles/etc. At least something to prevent it from interfering with 3rd-party entities, until wire-based entity check is added.
```

Comment:
again with the "interfering with 3rd party entities" except it doesn't and can't interfere in any way whatsoever